src/Compilers/Core/Portable/SourceGeneration/IncrementalContexts.cs

```diff
@@ -209,19 +252,19 @@ internal readonly struct IncrementalExecutionContext
 
         internal readonly GeneratorRunStateTable.Builder GeneratorRunStateBuilder;
 
-        internal readonly ArrayBuilder<(string Key, string Value)> HostOutputBuilder;
+        internal readonly ImmutableDictionary<string, object>.Builder HostOutputBuilder;
 
         public IncrementalExecutionContext(DriverStateTable.Builder? tableBuilder, GeneratorRunStateTable.Builder generatorRunStateBuilder, AdditionalSourcesCollection sources)
         {
             TableBuilder = tableBuilder;
             GeneratorRunStateBuilder = generatorRunStateBuilder;
             Sources = sources;
-            HostOutputBuilder = ArrayBuilder<(string, string)>.GetInstance();
+            HostOutputBuilder = ImmutableDictionary.CreateBuilder<string, object>();
             Diagnostics = DiagnosticBag.GetInstance();
         }
 
-        internal (ImmutableArray<GeneratedSourceText> sources, ImmutableArray<Diagnostic> diagnostics, GeneratorRunStateTable executedSteps, ImmutableArray<(string Key, string Value)> hostOutputs) ToImmutableAndFree()
-                => (Sources.ToImmutableAndFree(), Diagnostics.ToReadOnlyAndFree(), GeneratorRunStateBuilder.ToImmutableAndFree(), HostOutputBuilder.ToImmutableAndFree());
+        internal (ImmutableArray<GeneratedSourceText> sources, ImmutableArray<Diagnostic> diagnostics, GeneratorRunStateTable executedSteps, ImmutableDictionary<string, object> hostOutputs) ToImmutableAndFree()
+                => (Sources.ToImmutableAndFree(), Diagnostics.ToReadOnlyAndFree(), GeneratorRunStateBuilder.ToImmutableAndFree(), HostOutputBuilder.ToImmutable());
```

Comment:
Why don't we free the HostOutputBuilder here?

Reply:
Because its not pooled, so there is nothing to free.